### DISCUSSION TITLE
ImageResponseのキャッシュの設定方法について

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#135] [Q&A] [answered]
以下のとおり、ImageResponseのオプションにはheadersがあります

```ts
import { ImageResponse } from 'next/og'
 
new ImageResponse(
  element: ReactElement,
  options: {
    width?: number = 1200
    height?: number = 630
    emoji?: 'twemoji' | 'blobmoji' | 'noto' | 'openmoji' = 'twemoji',
    fonts?: {
      name: string,
      data: ArrayBuffer,
      weight: number,
      style: 'normal' | 'italic'
    }[]
    debug?: boolean = false
 
    // Options that will be passed to the HTTP response
    status?: number = 200
    statusText?: string
    headers?: Record<string, string>
  },
)
```

こちらで設定するのかな？と思い、試しに以下のように設定してみました

```ts
headers: { "Cache-Control": "public, max-age=315360000, immutable" },
```

そして、開発者ツールのネットワークタブを開きながら実行。
レスポンスヘッダーは以下の通りです

```ts
cache-control:
no-cache, no-store, public, max-age=315360000, immutable
```

`no-cache, no-store`がデフォルトで設定されている？

キャッシュの設定はImageResponseのheadersではせず、fetchするときにするものなのでしょうか？それとも、もっと別のところ……？

**Comment by kagomen** [comment]
`npm run preview`してリクエストを実行してみたら、キャッシュコントロールが以下のようになっていました！

```
cache-control:
public, immutable, no-transform, max-age=31536000,
```

さらに、現在の本番環境では以下の通りでした
```
cache-control:
public, immutable, no-transform, max-age=31536000
```


> public: このリソースは公開されており、共有キャッシュ（プロキシサーバーやCDNなど）でキャッシュできることを示します。
immutable: リソースの内容が変更されないことを示します。これにより、ブラウザは有効期限内のリソースを再検証せずに使用できます。
no-transform: 中間キャッシュやプロキシがリソースを変換（例：画像の圧縮）してはいけないことを示します。
max-age=31536000: リソースが31,536,000秒（1年）の間有効であることを示します。この期間中、クライアントは新しいバージョンをチェックせずにキャッシュされたバージョンを使用できます。
この設定の効果：
長期キャッシュ: リソースは1年間キャッシュされ、この期間中はネットワークリクエストを節約できます。
効率的な配信: CDNやプロキシサーバーでキャッシュできるため、配信が高速化されます。
不変性の保証: immutable フラグにより、有効期限内の不必要な再検証を防ぎます。
元のフォーマット保持: no-transform により、中間者による意図しない変更を防ぎます。
この設定は、頻繁に変更されない静的アセット（画像、CSS、JavaScriptファイルなど）に適しています。ただし、コンテンツが更新された場合、新しいURLを使用するなどの戦略が必要になります（例：ファイル名にハッシュを含める）。


特に設定する必要はなさ、そう……？

**Reply by kazzyfrog** [reply, marked as answer]
そしたら、
ユーザーエージェント（ブラウザ）のキャッシュは、そのままで良さそうですね！！

ただ、ユーザーAさん、Bさんがいたとして、
- 8/27に、Aさんが、`?theme=abc&emoji=🫠`にアクセスした場合、次回以降、Aさんの同じアクセスはキャッシュが利用される
- 8/28に、 Bさんが、`?theme=abc&emoji=🫠`にアクセスした場合は、サーバーにリクエストが飛ぶ

なので、理想は、サーバーでのデータキャッシュができたら良さそうではありますよね！

100人が、`?theme=abc&emoji=🫠`にアクセスした場合でも、１回のデータを利用できるのが、、理想？

多分、fetchでデータを取得すれば、デプロイ後も永続的なデータキャッシュを保持できるっぽいんですけど、、
最初は、そこまでキャッシュ気にしなくてもいいかもですね！w
（APIに、さらにfetchを噛ませるの大変そうなので。。）

**Reply by kazzyfrog** [reply]
参考：
> ルートハンドラーはデフォルトでキャッシュされない

公式：https://nextjs.org/docs/app/building-your-application/routing/route-handlers#caching

---

> [unstable_cache](https://nextjs.org/docs/app/api-reference/functions/unstable_cache) というベータ機能を使うことで ORM でもデータキャッシュ可能になりますが、不安定でバグが多いためまだ使える段階にありません。

https://zenn.dev/nino/books/30e21d37af73b5/viewer/cache

---

https://zenn.dev/cybozu_frontend/articles/next-caching-fetch-cache

なんか、めっちゃ複雑な上に、安定していない感じがしますよね！

**Reply by kagomen** [reply]
@kazzyfrog 
なるほどー……？
奥が深すぎて、理解が難しい……🫨

UIコンポーネント（クライアイント側）で `/api/v1/lgtm-images?theme=abc`に`fetch`する時にキャッシュの指定をするもんだと思っていたのですが、それとはまた別に、API側でさらに`fetch`するというテクもあるんですね……？🤯

**Reply by kazzyfrog** [reply]
> UIコンポーネント（クライアイント側）で /api/v1/lgtm-images?theme=abcにfetchする時にキャッシュの指定をする

自分たちのサイト上では、おっしゃる通りです！
ただ、実際ユーザーはAPIの方のURLをGitHubに入れると思うので、fetchを介さないんですよね。。！

> それとはまた別に、API側でさらにfetchするというテクもあるんですね……？
なので、多分現状で、データキャッシュを使おうとしたら、そうなるのかなと思います！

今実装している画像生成用APIエンドポイントと、
それを叩いて配信するための、fetch用エンドポイントを立てれば、多分データキャッシュを実装できるのかな？と！🧐

**Reply by kagomen** [reply]
> 実際ユーザーはAPIの方のURLをGitHubに入れると思うので、fetchを介さない

なるほど、そういうことか……！

> それを叩いて配信するための、fetch用エンドポイントを立てれば、多分データキャッシュを実装できるのかな？

なんとなく方針がわかった感じがします！解説ありがとうございますー！🙏🙏✨

**Comment by kazzyfrog** [comment]
調べてみたら、まだまだキャッシュ周り安定していないっぽいですね。。！

**Reply by kagomen** [reply]
そうなんですね
一旦、設定しない方向で、良いのでしょうか……？👀

**Reply by kazzyfrog** [reply]
一旦設定しない方向でも、いい気がしてます！！

- リクエスト数の制限とか、気にしないといけなくなったら 
- or キャッシュ周り深掘りしたくなったら 

で、いいかもですねw

**Reply by kagomen** [reply]
承知しました！
そしたらイシューはそのままにしておいて、

- assignはずす
- priority lowラベルをつける
- このdiscussionsをリンクする

を、しておきます！👍

**Reply by kazzyfrog** [reply]
助かりますー！！👍
